jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -554,27 +554,63 @@ bool
 ecma_string_get_array_index (const ecma_string_t *str_p, /**< ecma-string */
                              uint32_t *out_index_p) /**< [out] index */
 {
-  bool is_array_index = true;
-  if (ECMA_STRING_GET_CONTAINER (str_p) == ECMA_STRING_CONTAINER_UINT32_IN_DESC)
+  uint32_t index;
+  const ecma_string_container_t type = ECMA_STRING_GET_CONTAINER (str_p);
+
+  if (type == ECMA_STRING_CONTAINER_UINT32_IN_DESC)
+  {
+    index = str_p->u.uint32_number;
+  }
+  else if (type == ECMA_STRING_CONTAINER_MAGIC_STRING)
   {
-    *out_index_p = str_p->u.uint32_number;
+    return false;
   }
   else
   {
-    ecma_number_t num = ecma_string_to_number (str_p);
-    *out_index_p = ecma_number_to_uint32 (num);
+    lit_utf8_size_t size;
+    const lit_utf8_byte_t *raw_str_p;
 
-    ecma_string_t *to_uint32_to_string_p = ecma_new_ecma_string_from_uint32 (*out_index_p);
+    if (unlikely (type == ECMA_STRING_CONTAINER_MAGIC_STRING_EX))
+    {
+      size = lit_get_magic_string_ex_size (str_p->u.magic_string_ex_id);
+      raw_str_p = lit_get_magic_string_ex_utf8 (str_p->u.magic_string_ex_id);
+    }
+    else
+    {
+      JERRY_ASSERT (type == ECMA_STRING_CONTAINER_HEAP_UTF8_STRING);
 
-    is_array_index = ecma_compare_ecma_strings (str_p,
-                                                to_uint32_to_string_p);
+      size = str_p->u.utf8_string.size;
+      raw_str_p = (const lit_utf8_byte_t *) (str_p + 1);
+    }
 
-    ecma_deref_ecma_string (to_uint32_to_string_p);
-  }
+    if (size > 10
+        || (*raw_str_p == LIT_CHAR_0 && size > 1))
+    {
+      return false;
+    }
 
-  is_array_index = is_array_index && (*out_index_p != ECMA_MAX_VALUE_OF_VALID_ARRAY_INDEX);
+    index = 0;
+    for (lit_utf8_size_t i = 0; i < size; i++)
+    {
+      if (raw_str_p[i] > LIT_CHAR_9 || raw_str_p[i] < LIT_CHAR_0)
+      {
+        return false;
+      }
+
+      const uint32_t tmp = (index * 10) + (uint32_t) (raw_str_p[i] - LIT_CHAR_0);
+
+      if (unlikely (tmp < index))
```

Comment:
Such constructs are usually optimized out by compilers. I had similar bugs before.

I would do something like this:
1) end =size == 10 ? 9 : size;
2) the usual for loop [0..end] -- we know the number fits in index.
3) if size == 10, then do the appropriate checks. This is really a rare case.
 if (index > UINT32_MAX / 10) return false;
 if (index == UINT32_MAX / 10) -- another check
 else return true;